R/req-url.R

```diff
@@ -46,15 +50,79 @@ req_url <- function(req, url) {
 
 #' @export
 #' @rdname req_url
-req_url_query <- function(.req, ...) {
+#' @param .multi Controls what happens when an element of `...` is a vector
+#'   containing multiple values:
+#'
+#'   * `"error"`, the default, throws an error.
+#'   * `"comma"`, separates values with a `,`, e.g. `?x=1,2`.
+#'   * `"pipe"`, separates values with a `|`, e.g. `?x=1|2`.
+#'   * `"explode"`, turns each element into its own parameter, e.g. `?x=1&x=2`.
+#'
+#'   If none of these functions work, you can alternatively supply a function
+#'   that takes a character vector and returns a string.
+req_url_query <- function(.req,
+                          ...,
+                          .multi = c("error", "comma", "pipe", "explode")) {
   check_request(.req)
+  if (is.function(.multi)) {
+    multi <- .multi
+  } else {
+    multi <- arg_match(.multi)
+  }
+
+  dots <- list2(...)
+
+  type_ok <- map_lgl(dots, function(x) is_atomic(x) || is.null(x))
+  if (any(!type_ok)) {
+    cli::cli_abort(
+      "All elements of {.code ...} must be either an atomic vector or NULL."
+    )
+  }
+
+  n <- lengths(dots)
+  if (any(n > 1)) {
+    if (is.function(multi)) {
+      dots[n > 1] <- lapply(dots[n > 1], format_query_param)
+      dots[n > 1] <- lapply(dots[n > 1], multi)
+      dots[n > 1] <- lapply(dots[n > 1], I)
+    } else if (multi == "comma") {
+      dots[n > 1] <- lapply(dots[n > 1], format_query_param)
+      dots[n > 1] <- lapply(dots[n > 1], paste0, collapse = ",")
+      dots[n > 1] <- lapply(dots[n > 1], I)
+    } else if (multi == "pipe") {
+      dots[n > 1] <- lapply(dots[n > 1], format_query_param)
+      dots[n > 1] <- lapply(dots[n > 1], paste0, collapse = "|")
+      dots[n > 1] <- lapply(dots[n > 1], I)
+    } else if (multi == "explode") {
+      dots <- explode(dots)
+    } else if (multi == "error") {
+      cli::cli_abort(c(
+        "All vector elements of {.code ...} must be length 1.",
+        i = "Use {.arg multi} to choose a strategy for handling."
```

Comment:
The info sentence feels unfinished (for handling what?). Maybe add something like "for handling multi element vectors".